src/lib/Live/Eval.hs

```diff
@@ -35,17 +39,25 @@ watchAndEvalFile :: FilePath -> EvalConfig -> TopStateEx
 watchAndEvalFile fname opts env = do
   (_, resultsChan) <- spawn logServer
   let cfg = (opts, subChan Publish resultsChan)
-  (_, sourceChan) <- spawn $ runDriver cfg env
-  forkWatchFile fname sourceChan
+  (_, driverChan) <- spawn $ runDriver cfg env
+  forkWatchFile fname $ subChan FileChanged driverChan
   return $ subChan Subscribe resultsChan
 
 -- === executing blocks concurrently ===
 
 type SourceContents = String
 
 type DriverCfg = (EvalConfig, PChan RFragment)
-type DriverState = (WithId TopStateEx, CacheState)
-type DriverM = ReaderT DriverCfg (StateT DriverState IO)
+-- The evaluation-in-progress state is
+-- - The (identified) current top-level environment
+-- - The ThreadId of the currently running worker (if any)
+-- - The list of blocks that remain to be evaluated (if any)
+-- If the worker exists, it is working on evaluating the first element
+-- of the blocks list in the environment given by the TopStateEx.
+-- This is consistent at entry and exit from handling each message,
+-- but may be briefly inconsistent mid-handler.
```

Comment:
This consistency comment makes me worried. No other thread can notice this, right? The source evaluation state is entirely managed by the coordinator thread?

Reply:
Correct.  This is the actor model -- the only cross-thread information passing is by explicitly sending messages on explicit channels.  The state type has no channels in it, so no other thread can affect it.  The driver also does not send its state in messages, so no other thread can detect the transient inconsistency.

Reply:
Ok then I am a little confused about the "mid-handler" comment. Can you clarify what the "handler" is?